### PR TITLE
add v1 masking

### PR DESCRIPTION
**Pull request by younesbelkada**
Supports tokens masking for text enviornments so that the generated tokens can be passed to the PPO Trainer
Now the api looks like the following:

```python
from trl import TextEnvironment, AutoModelForCausalLMWithValueHead, PPOTrainer, PPOConfig
from transformers import AutoTokenizer, load_tool

tool = load_tool("ybelkada/simple-calculator")

model_id = "gpt2"

model = AutoModelForCausalLMWithValueHead.from_pretrained(model_id).to(0)
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token


prompt = """\
What is 12.1 + 1 - 3?
<request><SimpleCalculatorTool>12.1 + 1<call>13.1<response>
<request><SimpleCalculatorTool>13.1 - 3<call>10.1<response>
Result = 10.1 <submit>

"""

reward = lambda x: 1

env = TextEnvironment(model, tokenizer, [tool], reward, prompt, generation_kwargs={"max_new_tokens": 32})

ppo_config = PPOConfig(
    batch_size=2,
    mini_batch_size=1,
)
ppo_trainer = PPOTrainer(ppo_config, model, None, tokenizer)

queries, responses, masks, rewards, history = env.run(["What is 387 * 228?", "What is 23.1 + 1 - 3?"])
ppo_trainer.step(queries, responses, rewards, masks)
```

### CODE REVIEW

**Review by vwxyzjn — COMMENTED** [review]
Very nice work!! Some quick comments

File: trl/environment/base.py [hunk 1]
```diff
@@ -99,6 +104,30 @@ def __init__(self, model, tokenizer, tools, reward_fn, prompt, max_turns=4, gene
...
+        return [self.request_token, self.call_token, self.response_token, self.submit_token] + self.tool_tokens
+
+    def encode_prompt(self, prompt, add_eos_token=False):
+        if not all([extra_token in self.tokenizer.special_tokens_map for extra_token in self.extra_tokens]):
+            logging.warning("Adding extra tokens to tokenizer - this will change the tokenizer.")
+            self.prompt_tokenizer.add_special_tokens({"additional_special_tokens": self.extra_tokens})
+        if self.prompt_tokenizer.pad_token_id is None:
+            raise ValueError("Tokenizer must have a pad token.")
+        if add_eos_token:
+            prompt = prompt + self.prompt_tokenizer.eos_token
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/environment/base.py:118]: This is to be consistent with the default behavior of the tokenizers. 
I need to dig more but for some reason the copied tokenizer did not added the EOS token after encoding the text. We may not need that at the end after thinking a bit

File: trl/environment/base.py [hunk 2]
```diff
@@ -99,6 +104,30 @@ def __init__(self, model, tokenizer, tools, reward_fn, prompt, max_turns=4, gene
...
+
+        encoded_prompt = self.prompt_tokenizer(prompt, return_tensors="pt")["input_ids"]
+        encoded_prompt = encoded_prompt.to(self.current_device)
+
+        attention_mask = (encoded_prompt < self.prompt_tokenizer.vocab_size) * 1.0
+        encoded_prompt[~(attention_mask.bool())] = self.prompt_tokenizer.pad_token_id
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/environment/base.py]: 1/ Yes just added your suggestion ! makes sense!
2/ This is done under the hood by the model I think, if you explicitly pass attention masks to the model as we do, it should automatically set the attention weights of these tokens to -inf

File: trl/environment/base.py [hunk 3]
```diff
@@ -75,9 +76,10 @@ def show(self):
 
 
 class TextEnvironment:
-    def __init__(self, model, tokenizer, tools, reward_fn, prompt, max_turns=4, generation_kwargs=None):
+    def __init__(self, model, tokenizer, tools, reward_fn, prompt, tool_tokens, max_turns=4, generation_kwargs=None):
```
Comment (lvwerra) [inline trl/environment/base.py:79]: we don't need this anymore, right?

File: trl/environment/base.py [hunk 4]
```diff
@@ -75,9 +76,10 @@ def show(self):
...
         self.model = model
         self.tokenizer = tokenizer
+        self.prompt_tokenizer = deepcopy(tokenizer)
```
Comment (lvwerra) [inline trl/environment/base.py:82]: also not needed

File: trl/environment/base.py [hunk 5]
```diff
@@ -89,32 +91,78 @@ def __init__(self, model, tokenizer, tools, reward_fn, prompt, max_turns=4, gene
...
-        self.compute_reward(histories)
+        histories = self.compute_reward(histories)
 
-        return histories
+        for i, history in enumerate(histories):
+            histories_rewards.append(torch.Tensor([history.reward]))
+            histories_responses[i] = torch.cat(histories_responses[i], dim=0).squeeze()
+            responses_masks[i] = torch.cat(responses_masks[i], dim=0).squeeze()
+
+        return (histories_queries, histories_responses), (responses_masks, queries_masks), histories_rewards, histories
```
Comment (lvwerra) [inline trl/environment/base.py:165]: Still not a big fan of the nested returns, I think this will make it not very user friendly.

File: trl/trainer/ppo_config.py [hunk 6]
```diff
@@ -146,6 +146,10 @@ class PPOConfig(object):
         default=1,
         metadata={"help": "Number of steps between comparison of the current reward with the best seen so far"},
     )
+    use_text_environment: Optional[bool] = field(
+        default=False,
+        metadata={"help": "Whehter to use the text environment or not"},
+    )
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/trainer/ppo_config.py]: Yeah makes sense!

File: trl/environment/base.py [hunk 7]
```diff
@@ -89,32 +91,78 @@ def __init__(self, model, tokenizer, tools, reward_fn, prompt, max_turns=4, gene
...
+    def encode_history(self, history, is_system_prompt=False):
+        encoded_prompt = self.tokenizer(history.last_text_segment, return_tensors="pt").to(self.current_device)
+        input_ids = encoded_prompt["input_ids"]
+
+        if is_system_prompt:
+            attention_mask = torch.zeros_like(input_ids)
+        else:
+            attention_mask = torch.ones_like(input_ids)
+
+        return {"input_ids": input_ids, "attention_mask": attention_mask}
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/environment/base.py:112]: Hmm I am afraid this might add a lot of hacks and boilerplate in the TextHistory class, wdyt about leaving it as it is for now?

[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/environment/base.py:112]: Currently the encode_history method is rather simple

[the remark this answers is not in the document]
Comment (younesbelkada) [reply trl/environment/base.py:112]: Or maybe I just need a clearer picture of the methods you had in mind for that

**Comment by vwxyzjn** [comment]
image

The learning is clearly promising but somehow collapsed... See report here https://wandb.ai/costa-huang/cleanrl/reports/Regression-Report-wandb--Vmlldzo0NzA3MzEy
